### PR TITLE
[spiflash] Add support to override the FTDI USB device ID

### PR DESCRIPTION
**Pull request by moidx**
This PR adds support for overriding the FTDI device ID in spiflash. This is required when working with FTDI MPSSE USB cables or other FPGA boards.

The following command line options are added to spiflash:
  
* `--dev_id`: FTDI USB vendor and product ID.
* `--dev_sn`: FTDI USB serial number.

Other changes:

* Convert documentation to Doxygen format. Simple format conversion.
* Update README.md with instructions on how to override the device ID.

### CODE REVIEW

File: sw/host/spiflash/README.md [hunk 1]
```diff
@@ -73,6 +61,15 @@ If there are two FPGAs or multiple valid targets attached at the same time, it i
 
 ```console
 $ cd ${REPO_TOP}
-$ build-bin/sw/host/spiflash/spiflash \
-  --input build-bin/sw/device/examples/hello_world/hello_world_fpga_nexysvideo.bin
+$ build-out/sw/host/spiflash/spiflash --input ${FLASH_BIN}
```
Comment (lenary) [inline sw/host/spiflash/README.md:64]: We should keep using the `build-bin` versions of the binary, I think.

File: sw/host/spiflash/spiflash.cc [hunk 2]
```diff
@@ -24,22 +23,35 @@ using opentitan::spiflash::VerilatorSpiInterface;
 
 constexpr char kUsageString[] = R"R( usage options:
   --input=Input image in binary format.
+
+FTDI Options:
+  [--dev_id="vid:pid"] FTDI device ID.
+    vid: Vendor ID in string hex format.
+    pid: Product ID in string hext format.
```
Comment (lenary) [inline sw/host/spiflash/spiflash.cc:30]:
```suggestion
    pid: Product ID in string hex format.
```

File: sw/host/spiflash/ftdi_spi_interface.cc [hunk 3]
```diff
@@ -88,7 +78,24 @@ FtdiSpiInterface::~FtdiSpiInterface() {
...
+
+  struct mpsse_context *ctx = nullptr;
+  if (options_.device_vendor_id == 0 || options_.device_product_id == 0) {
+    ctx = MPSSE(SPI0, options_.spi_frequency, MSB);
+  } else {
+    const char *serial_number = options_.device_serial_number.empty()
+                                    ? nullptr
+                                    : options_.device_serial_number.c_str();
+    ctx = Open(options_.device_vendor_id, options_.device_product_id, SPI0,
+               options_.spi_frequency, MSB, IFACE_A, nullptr, serial_number);
```
Comment (mcy) [inline sw/host/spiflash/ftdi_spi_interface.cc:102]: Would you mind adding /* arg = */ annotations to function arguments here that aren't obvious? E.g. SPI0, MSB, etc.

File: sw/host/spiflash/ftdi_spi_interface.h [hunk 4]
```diff
@@ -23,7 +23,29 @@ struct MpsseHandle;
 // This class is not thread safe.
 class FtdiSpiInterface : public SpiInterface {
  public:
-  FtdiSpiInterface();
+  // FTDI SPI configuration options.
+  struct Options {
+    // USB device vendor ID.
+    int device_vendor_id;
```
Comment (mcy) [inline sw/host/spiflash/ftdi_spi_interface.h:31]: Consider switching `int` to `int32_t`.

File: sw/host/spiflash/verilator_spi_interface.h [hunk 5]
```diff
@@ -12,26 +12,27 @@
...
       : spi_filename_(spi_filename), fd_(-1) {}
 
-  // Closes the internal file handle used to communicate with the SPI device.
+  /**
+   * Closes the internal file handle used to communicate with the SPI device.
+   */
   ~VerilatorSpiInterface() override;
 
-  // Initialize interface.
+  // SpiInterface functions.
```
Comment (mcy) [inline sw/host/spiflash/verilator_spi_interface.h:33]: Drop this comment. It merely says what, but not why.

File: sw/host/spiflash/spiflash.cc [hunk 6]
```diff
@@ -55,23 +67,55 @@ bool GetFileContents(const std::string &filename, std::string *contents) {
...
+
+/*
+ * Parse command line arguments and store results in `options`.
+ */
 bool ParseArgs(int argc, char **argv, SpiFlashOpts *options) {
   assert(options);
   const struct option long_options[] = {
       {"input", required_argument, nullptr, 'i'},
+      {"dev_id", required_argument, nullptr, 'd'},
+      {"dev_sn", required_argument, nullptr, 'n'},
```
Comment (imphil) [inline sw/host/spiflash/spiflash.cc:112]: Nit: Personally, I'd prefer `-` in arguments over `_`, so `dev-id` and `dev-sn`

File: sw/host/spiflash/README.md (hunk 1 above)
Comment (moidx) [reply sw/host/spiflash/README.md]: Done.

File: sw/host/spiflash/spiflash.cc (hunk 2 above)
Comment (moidx) [reply sw/host/spiflash/spiflash.cc]: Done.

File: sw/host/spiflash/ftdi_spi_interface.cc (hunk 3 above)
Comment (moidx) [reply sw/host/spiflash/ftdi_spi_interface.cc]: Done.

File: sw/host/spiflash/ftdi_spi_interface.h (hunk 4 above)
Comment (moidx) [reply sw/host/spiflash/ftdi_spi_interface.h]: Done.

File: sw/host/spiflash/verilator_spi_interface.h (hunk 5 above)
Comment (moidx) [reply sw/host/spiflash/verilator_spi_interface.h]: Done.

File: sw/host/spiflash/spiflash.cc (hunk 6 above)
Comment (moidx) [reply sw/host/spiflash/spiflash.cc]: Done.

**Comment by alistair23** [comment]
This PR (specifically commit [spiflash] Add support for configurable FTDI device IDs) breaks the spiflash program for me.

When flashing a binary to the Nexys FPGA before this commit I could run:

```shell
build-out/sw/host/spiflash/spiflash --input=build-out/sw/device/examples/hello_world/hello_world_fpga_nexysvideo.bin
```

After this PR I just get this error instead:

```
Unable to open FTDI SPI interface.
```

I raised an issue here: https://github.com/lowRISC/opentitan/issues/3315

**Pull request merged**